internal/gatewayapi/securitypolicy.go

```diff
@@ -388,21 +379,37 @@ func (t *Translator) translateSecurityPolicyForRoute(
 	// Note: there are multiple features in a security policy, even if some of them
 	// are invalid, we still want to apply the valid ones.
 	prefix := irRoutePrefix(route)
-	for _, x := range xdsIR {
-		for _, h := range x.HTTP {
-			for _, r := range h.Routes {
-				// Apply if there is a match
-				// route is associated with a Gateway API xRoute
-				if strings.HasPrefix(r.Name, prefix) {
-					// This security policy matches the current route.
-					// It should only be accepted if it doesn't match any other route
-					r.Security = &ir.SecurityFeatures{
-						CORS:          cors,
-						JWT:           jwt,
-						OIDC:          oidc,
-						BasicAuth:     basicAuth,
-						ExtAuth:       extAuth,
-						Authorization: authorization,
+	parentRefs := GetParentReferences(route)
+	for _, p := range parentRefs {
+		parentRefCtx := GetRouteParentContext(route, p)
+		gtwCtx := parentRefCtx.GetGateway()
+		if gtwCtx != nil {
+			if policy.Spec.ExtAuth != nil {
```

Comment:
merge these two conditions, like `gtwCtx != nil && policy.Spec.ExtAuth != nil`. there're too much code blocks in this loop

Reply:
we still need to loop over listeners to set other `SecurityFeatures` even if `policy.Spec.ExtAuth == nil`. I have refactored the code to `continue` early for reducing nesting.